test/functional/p2p_private_broadcast.py

```diff
@@ -0,0 +1,307 @@
+            # may be active (B), or may have already been closed (C).
+            peer = self.socks5_server.conf.destinations[i]["node"]
+            peer.wait_until(lambda: peer.message_count["version"] == 1, check_connected=False)
+            # Now it is either (B) or (C).
+            requested_to_addr = self.socks5_server.conf.destinations[i]["requested_to_addr"]
+            if peer.last_message["version"].nServices != 0:
+                self.log.debug(f"Outbound connection i={i} to {requested_to_addr} not a private broadcast, ignoring it (maybe feeler or extra block only)")
+                continue
+            self.log.debug(f"Outbound connection i={i} to {requested_to_addr} must be a private broadcast, checking it")
+            peer.wait_for_disconnect()
+            # Now it is (C).
+            assert_equal(peer.message_count, {
+                "version": 1,
+                "verack": 1,
+                "tx": 1,
+                "ping": 1
+            })
+            dummy_address = CAddress()
+            dummy_address.nServices = 0
+            assert_equal(peer.last_message["version"].nVersion, P2P_VERSION)
+            assert_equal(peer.last_message["version"].nServices, 0)
+            assert_equal(peer.last_message["version"].nTime, 0)
+            assert_equal(peer.last_message["version"].addrTo, dummy_address)
+            assert_equal(peer.last_message["version"].addrFrom, dummy_address)
+            assert_equal(peer.last_message["version"].strSubVer, "/pynode:0.0.1/")
+            assert_equal(peer.last_message["version"].nStartingHeight, 0)
+            assert_equal(peer.last_message["version"].relay, 0)
+            assert_equal(peer.last_message["tx"].tx.rehash(), tx["txid"])
+            self.log.debug(f"Outbound connection i={i} is proper private broadcast, test ok")
+            num_private_broadcast_opened += 1
+            if num_private_broadcast_opened == NUM_PRIVATE_BROADCAST_PER_TX:
+                break
+        assert_equal(num_private_broadcast_opened, NUM_PRIVATE_BROADCAST_PER_TX)
+
+        wtxid_int = int(tx["wtxid"], 16)
+        inv = CInv(MSG_WTX, wtxid_int)
+
+        self.log.info("Checking that the transaction is not in the originator node's mempool")
+        assert_equal(len(tx_originator.getrawmempool()), 0)
+
+        self.log.info("Sending INV from an observer and waiting for GETDATA from node")
+        observer_inbound.tx_store[wtxid_int] = tx["tx"]
+        assert "getdata" not in observer_inbound.last_message
+        observer_inbound.send_message(msg_inv([inv]))
+        observer_inbound.wait_until(lambda: "getdata" in observer_inbound.last_message)
+        self.wait_until(lambda: len(tx_originator.getrawmempool()) > 0)
+
+        self.log.info("Waiting for normal broadcast to another observer")
+        observer_outbound = self.socks5_server.conf.destinations[0]["node"]
+        observer_outbound.wait_for_inv([inv])
```

Comment:
I think the test needs coverage for:

1) retrying broadcasts
2) sending the same tx again via sendrawtransaction
3) sending a different wtxid via sendrawtransaction
4) sending a tx with in-mempool dependencies (both that succeed and fail due to existing/missing parent)
5) rpc failing if it detects `!g_reachable_nets`

Reply:
> I think the test needs coverage for:

I was stuck at https://github.com/bitcoin/bitcoin/issues/30543, thank you @mzumsande for resolving it!

>     1. retrying broadcasts

Done.

>     2. sending the same tx again via sendrawtransaction

Done.

>     3. sending a different wtxid via sendrawtransaction

Done, but the new tx has garbage witness. Would be nice to add one more test with a valid but different witness.

>     4. sending a tx with in-mempool dependencies (both that succeed and fail due to existing/missing parent)

Done, both.

>     5. rpc failing if it detects `!g_reachable_nets`

Done.

>     6. what happens if there's not enough private connections when first broadcast attempt is tried, but more found later

Private broadcast connections are created on demand. When first broadcast attempt is tried they are 0 (I did not add new test).

>     7. what happens to pending broadcasts on restart of node

They are lost. Persistence over restart is not in this PR, but will be in a subsequent one (I did not add new test).

Reply:
thanks for attending to this laundry list :+1: